indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -305,6 +346,125 @@ public TaskLock lock(final Task task, final Interval interval) throws Interrupte
   }
 
   /**
+   * Check priorities of existing tasks, if all has lesser priorities and no lock is non-preemptive then return true
+   * If this task can acquire lock and some lower priority TaskLocks need to be revoked then populate tasksToPreempt Map with the entry - taskId and TaskLock
+   * */
+  private Pair<Boolean, Map<String, TaskLock>> canAcquireLock(final List<TaskLockPosse> foundPosses, int priority) {
+    boolean canAcquire = true;
+    final Map<String, TaskLock> tasksToPreempt = new HashMap<>();
+    for (TaskLockPosse taskLockPosse : foundPosses) {
+      if(priority <= taskLockPosse.getTaskLock().getPriority() || !taskLockPosse.getTaskLock().isPreemptive()) {
+        // Check whether the tasks (in the taskLockPosse) which are blocking the new task are actually running,
+        // if not remove their taskLocks and info from taskStorage and taskLockPosse
+        for (String taskId: taskLockPosse.getTaskIds()) {
+          if (!isTaskAlive(taskId)) {
+            Optional<Task> taskToUnlock = taskStorage.getTask(taskId);
+            if (taskToUnlock.isPresent()) {
+              remove(taskToUnlock.get());
+            } else {
+              log.warn("Got task [%s] with an active TaskLock but is not present in Metadata Storage !!", taskId);
+            }
+          } else {
+            canAcquire = false;
+            tasksToPreempt.clear();
+          }
+        }
+      } else {
+        // This task has higher priority than existing tasks, unlock the existing task lock
+        for (String taskId : taskLockPosse.getTaskIds()) {
+          tasksToPreempt.put(taskId, taskLockPosse.getTaskLock());
+        }
+      }
+    }
+    return new Pair<>(canAcquire, tasksToPreempt);
+  }
+
+  /**
+   * Handle lower priority locks. As of now it just unlocks the TaskLocks
```

Comment:
As of now handle just means revoke the tasklocks of lower priority tasks. I can rename the method name to 'revokeLowerPriorityTaskLocks' to be more clear